pkg/reporter/gitlab.go

```diff
@@ -0,0 +1,358 @@
+	// NOT GITLAB BUT WE ARE USING THIS FOR OUR CUSTOM IDENTIFIER
+	gitLabIdentifierTypeMALWARE gitLabIdentifierType = "malware"
+)
+
+// gitLabIdentifier represents identifier information
+type gitLabIdentifier struct {
+	Type  gitLabIdentifierType `json:"type"`
+	Name  string               `json:"name"`
+	Value string               `json:"value"`
+	URL   string               `json:"url"`
+}
+
+// Severity represents severity of a vulnerability or malware
+type Severity string
+
+const (
+	SeverityUnknown  Severity = "Unknown"
+	SeverityCritical Severity = "Critical"
+	SeverityHigh     Severity = "High"
+	SeverityMedium   Severity = "Medium"
+	SeverityLow      Severity = "Low"
+)
+
+// gitLabVulnerability represents a vulnerability in GitLab format
+// Docs: https://docs.gitlab.com/development/integrations/secure/#vulnerabilities
+type gitLabVulnerability struct {
+	ID          string             `json:"id"`
+	Name        string             `json:"name"`
+	Description string             `json:"description"`
+	Severity    Severity           `json:"severity"`
+	Location    gitLabLocation     `json:"location"`
+	Identifiers []gitLabIdentifier `json:"identifiers"`
+}
+
+// gitLabScan represents scan information
+type gitLabScan struct {
+	Scanner   gitLabScanner `json:"scanner"`
+	Analyzer  gitLabScanner `json:"analyzer"` // Reusing GitLabScanner as they have same structure
+	Type      string        `json:"type"`
+	StartTime string        `json:"start_time"`
+	EndTime   string        `json:"end_time"`
+	Status    string        `json:"status"`
+}
+
+// gitLabReport represents the complete GitLab report currently using the 15.2.1 schema
+// and `dependency_scanning` type.
+// but can be extended to support other types and schemas in the future.
+// docs: https://docs.gitlab.com/development/integrations/secure/#report
+type gitLabReport struct {
+	Schema          string                `json:"schema"`
+	Version         string                `json:"version"`
+	Scan            gitLabScan            `json:"scan"`
+	Vulnerabilities []gitLabVulnerability `json:"vulnerabilities"`
+}
+
+type gitLabReporter struct {
+	config          GitLabReporterConfig
+	vulnerabilities []gitLabVulnerability
+	startTime       time.Time
+}
+
+// Ensure gitLabReporter implements Reporter interface
+var _ Reporter = (*gitLabReporter)(nil)
+
+func NewGitLabReporter(config GitLabReporterConfig) (Reporter, error) {
+	return &gitLabReporter{
+		config:          config,
+		vulnerabilities: make([]gitLabVulnerability, 0),
+		startTime:       time.Now(),
+	}, nil
+}
+
+func (r *gitLabReporter) Name() string {
+	return "GitLab Dependency Scanning Report Generator"
+}
+
+// GitLab requires time to be in pattern
+// "^\\d{4}-\\d{2}-\\d{2}T\\d{2}:\\d{2}:\\d{2}$"
+// Example: "2020-01-28T03:26:02"
+//
+// Docs (Schema Reference): https://gitlab.com/gitlab-org/security-products/security-report-schemas/-/blob/master/dist/sast-report-format.json#L497
+func gitlabFormatTime(t time.Time) string {
```

Comment:
If this function exists ONLY for the gitlab reporter then we should scope it within the `*gitLabReporter` struct I think?